File: Advanced/Ripples/WebGL/AnimatedRipples.js
// Cette classe permet d'appliquer une déformation type tremblement sur un FBO
// voir https://github.com/SFML/SFML/wiki/Source:-HeatHazeShader

// superclasses et classes nécessaires
Requires("libs/Process/Process.js");


class AnimatedRipples extends Process
{
    /**
     * Constructeur
     */
    constructor()
    {
        super("AnimatedRipples");

        // charger la texture de distorsion
        this.m_Texture = new Texture2D("data/textures/distorsion.png", gl.LINEAR, gl.REPEAT);

        // compiler le shader
        this.compileShader();
    }


    /**
     * retourne le source du Fragment Shader
     */
    getFragmentShader()
    {
        return dedent
            `#version 300 es
            precision mediump float;
            uniform sampler2D ColorMap;
            uniform sampler2D DistorsionMap;

            uniform float time;
            uniform float strength;
            uniform float speed;

            in vec2 frgTexCoord;

            out vec4 glFragColor;

            void main()
            {
                // coordonnées dans la texture de distorsion
                vec2 texCoords = frgTexCoord + fract(vec2(0.0, time * speed));

                // accès à la texture, elle retourne un décalage
                vec2 offset = texture(DistorsionMap, texCoords).xy * 2.0 - 1.0;

                // conditionnement du décalage
                offset *= strength;

                // coordonnées dans l'image à déformer
                texCoords = frgTexCoord + offset;

                // accès à l'image déformée
                glFragColor = texture(ColorMap, texCoords);
            }`;
    }


    /**
     * détermine où sont les variables uniform spécifiques de ce traitement
     */
    findUniformLocations()
    {
        // appeler la méthode de la superclasse
        super.findUniformLocations();

        // déterminer où sont les variables uniform spécifiques
        this.m_DistorsionMapLoc = gl.getUniformLocation(this.m_ShaderId, "DistorsionMap");
        this.m_StrengthLoc      = gl.getUniformLocation(this.m_ShaderId, "strength");
        this.m_SpeedLoc         = gl.getUniformLocation(this.m_ShaderId, "speed");
        this.m_TimeLoc          = gl.getUniformLocation(this.m_ShaderId, "time");
    }


    /**
     * applique une distorsion à l'image du FBO
     * @param fbo : FBO contenant l'image à traiter
     * @param strength : importance de l'effet
     * @param speed : vitesse de l'effet
     */
    process(fbo, strength=0.003, speed=1.0)
    {
        // préparer le shader pour le traitement
        this.startProcess();

        // fournir le color buffer du FBO à traiter
        this.setTextureUnit(gl.TEXTURE0, this.m_ColorMapLoc, fbo.getColorBuffer());
        this.m_Texture.setTextureUnit(gl.TEXTURE1, this.m_DistorsionMapLoc);

        // fournir les paramètres du traitement
        gl.uniform1f(this.m_StrengthLoc, strength);
        gl.uniform1f(this.m_SpeedLoc,    speed);
        gl.uniform1f(this.m_TimeLoc,     Utils.Time);

        // dessiner un quadrilatère avec les quatre vertices
        gl.drawArrays(gl.TRIANGLE_FAN, 0, 4);

        // désactiver les textures
        this.setTextureUnit(gl.TEXTURE0);
        this.setTextureUnit(gl.TEXTURE1);

        // libérer les ressources
        this.endProcess();
    }


    /**
     * Cette méthode supprime les ressources allouées
     */
    destroy()
    {
        this.m_Texture.destroy();
        super.destroy();
    }
}

File: Advanced/Ripples/C++/AnimatedRipples.cpp
// Cette classe permet d'appliquer une déformation type tremblement sur un FBO

#include <iostream>
#include <math.h>

#include <utils.h>


#include <AnimatedRipples.h>


/**
 * Constructeur
 */
AnimatedRipples::AnimatedRipples():
    Process("AnimatedRipples")
{
    // charger la texture de distorsion
    m_Texture = new Texture2D("data/textures/distorsion.png", GL_LINEAR, GL_REPEAT);

    // compiler le shader
    compileShader();
}


/**
 * retourne le source du Fragment Shader
 */
std::string AnimatedRipples::getFragmentShader()
{
    return
        "#version 300 es\n"
        "precision mediump float;\n"
        "uniform sampler2D ColorMap;\n"
        "uniform sampler2D DistorsionMap;\n"
        "\n"
        "uniform float time;\n"
        "uniform float strength;\n"
        "uniform float speed;\n"
        "\n"
        "in vec2 frgTexCoord;\n"
        "out vec4 glFragColor;\n"
        "\n"
        "void main()\n"
        "{\n"
        "    // coordonnées dans la texture de distorsion\n"
        "    vec2 texCoords = frgTexCoord + fract(vec2(0.0, time * speed));\n"
        "\n"
        "    // accès à la texture, elle retourne un décalage\n"
        "    vec2 offset = texture(DistorsionMap, texCoords).xy * 2.0 - 1.0;\n"
        "\n"
        "    // conditionnement du décalage\n"
        "    offset *= strength;\n"
        "\n"
        "    // coordonnées dans l'image à déformer\n"
        "    texCoords = frgTexCoord + offset;\n"
        "\n"
        "    // accès à l'image déformée\n"
        "    glFragColor = texture(ColorMap, texCoords);\n"
        "}";
}


/**
 * détermine où sont les variables uniform spécifiques de ce traitement
 */
void AnimatedRipples::findUniformLocations()
{
    // appeler la méthode de la superclasse
    Process::findUniformLocations();

    // déterminer où sont les variables uniform spécifiques
    m_DistorsionMapLoc = glGetUniformLocation(m_ShaderId, "DistorsionMap");
    m_StrengthLoc      = glGetUniformLocation(m_ShaderId, "strength");
    m_SpeedLoc         = glGetUniformLocation(m_ShaderId, "speed");
    m_TimeLoc          = glGetUniformLocation(m_ShaderId, "time");
}


/**
 * applique une distorsion à l'image du FBO
 * @param fbo : FBO contenant l'image à traiter
 * @param strength : importance de l'effet
 * @param speed : vitesse de l'effet
 */
void AnimatedRipples::process(FrameBufferObject* fbo, float strength, float speed)
{
    // préparer le shader pour le traitement
    startProcess();

    // fournir le color buffer du FBO à traiter
    setTextureUnit(GL_TEXTURE0, m_ColorMapLoc, fbo->getColorBuffer());
    m_Texture->setTextureUnit(GL_TEXTURE1, m_DistorsionMapLoc);

    // fournir les paramètres du traitement
    glUniform1f(m_StrengthLoc, strength);
    glUniform1f(m_SpeedLoc,    speed);
    glUniform1f(m_TimeLoc,     Utils::Time);

    // dessiner un quadrilatère avec les quatre vertices
    glDrawArrays(GL_TRIANGLE_FAN, 0, 4);

    // désactiver les textures
    setTextureUnit(GL_TEXTURE0);
    setTextureUnit(GL_TEXTURE1);

    // libérer les ressources
    endProcess();
}


/**
 * Cette méthode supprime les ressources allouées
 */
AnimatedRipples::~AnimatedRipples()
{
    delete m_Texture;
}
